libs/partners/cohere/langchain_cohere/cohere_agent.py

```diff
@@ -0,0 +1,120 @@
+from langchain_core.prompts.chat import ChatPromptTemplate
+from langchain_core.pydantic_v1 import BaseModel
+from langchain_core.runnables import Runnable, RunnablePassthrough
+from langchain_core.runnables.base import RunnableLambda
+from langchain_core.tools import BaseTool
+
+
+def create_cohere_tools_agent(
+    llm: BaseLanguageModel, tools: Sequence[BaseTool], prompt: ChatPromptTemplate
+) -> Runnable:
+    def llm_with_tools(input: Dict) -> Runnable:
+        tool_results = input["tool_results"] if len(input["tool_results"]) > 0 else None
+        tools = input["tools"] if len(input["tools"]) > 0 else None
+        return RunnableLambda(lambda x: x["input"]) | llm.bind(
+            tools=tools, tool_results=tool_results
+        )
+
+    agent = (
+        RunnablePassthrough.assign(
+            # Intermediate steps are in tool results.
+            # Edit below to change the prompt parameters.
+            input=lambda x: prompt.format_messages(
+                input=x["input"], agent_scratchpad=[]
+            ),
+            tools=lambda x: format_to_cohere_tools(tools, x["intermediate_steps"]),
+            tool_results=lambda x: format_to_cohere_tools_messages(
+                x["intermediate_steps"]
+            ),
+        )
+        | llm_with_tools
+        | CohereToolsAgentOutputParser()
+    )
+    return agent
+
+
+def format_to_cohere_tools(
+    tools: Sequence[BaseTool],
+    intermediate_steps: Sequence[Tuple[AgentAction, str]],
+) -> List[Dict[str, Any]]:
+    if (
+        len(intermediate_steps) == 1
+        and intermediate_steps[0][0].tool == "directly_answer"
+    ):
+        return []
+    return [convert_to_cohere_tool(tool) for tool in tools]
+
+
+def format_to_cohere_tools_messages(
+    intermediate_steps: Sequence[Tuple[AgentAction, str]],
+) -> list:
+    """Convert (AgentAction, tool output) tuples into tool messages."""
+    if len(intermediate_steps) == 0:
+        return []
+    tool_results = []
+    for agent_action, observation in intermediate_steps:
+        if agent_action.tool == "directly_answer":
+            continue
+        tool_results.append(
+            {
+                "call": {
+                    "name": agent_action.tool,
+                    "parameters": agent_action.tool_input,
+                },
+                "outputs": [{"answer": observation}],
+            }
+        )
+
+    return tool_results
+
+
+def convert_to_cohere_tool(
+    tool: Union[Dict[str, Any], Type[BaseModel], Callable, BaseTool],
+) -> Dict[str, Any]:
+    """Convert a raw function/class to an Cohere tool."""
+    if isinstance(tool, BaseTool):
+        return {
+            "name": tool.name,
+            "description": tool.description,
+            "parameter_definitions": tool.args,
+        }
+    else:
+        raise ValueError(
+            f"Unsupported tool type {type(tool)}. Tools must be passed in"
+            f" as BaseTool"
+        )
+
+
+class CohereToolsAgentOutputParser(
```

Comment:
Matching the format above will allow ChatCohere to be used with `OpenAIToolsAgentOutputParser`!

Reply:
so we can delete this

Reply:
That sounds good, however, we'd like the flexibility (and naming) in the partner package. We're only 'exporting' `create_cohere_tools_agent`, and if there are plans to refactor some of the tools interfaces in `langchain` we could later update this. Would that work for you?

Reply:
Sounds good. Added an integration test to confirm interface matches at least, and it's failing :(

Could you get that passing, then it looks good

Reply:
sure thing!

Reply:
I think I fixed a couple of bugs, but I also wanted to double check that the behaviour is correct as I changed the tests because they looked wrong to me:

1) In the non-stream test I updated it to use `invoke` instead of `stream`
2) What's the expected kwargs.tool_calls value when tools are bound to the request but the model indicates to not call any tool? The tests previously had this as "tool_calls is not present", but I think it should be "tool_calls is an empty sequence". Happy to change this though!